### PR TITLE
Fix message retry mechanisms

### PR DESCRIPTION
**Pull request by PurpShell**
Hey everyone! This fixes the "Message missing in node" error! Please take a look!

Big thanks to my Sponsors, they allow me to to do work like this. Check out my Sponsors page and donate if you can, it means a lot!
[Support my work on GitHub Sponsors](https://github.com/sponsors/PurpShell)

### CODE REVIEW

**Review by canove — DISMISSED** [review]

File: src/Socket/messages-recv.ts [hunk 1]
```diff
@@ -156,9 +160,9 @@ export const makeMessagesRecvSocket = (config: SocketConfig) => {
 						{
 							tag: 'retry',
 							attrs: {
-								count: retryCount.toString(),
```
Comment (canove) [inline src/Socket/messages-recv.ts]: rever this .toString()

**Review by canove — DISMISSED** [review]

File: src/Socket/messages-recv.ts [hunk 2]
```diff
@@ -156,9 +160,9 @@ export const makeMessagesRecvSocket = (config: SocketConfig) => {
...
-								id: node.attrs.id,
-								t: node.attrs.t,
+								count: retryCount as any,
+								id: node.attrs.id.toString(),
+								t: node.attrs.t.toString(),
```
Comment (purpshell) [inline src/Socket/messages-recv.ts]:
```suggestion
								count: retryCount.toString(),
								id: node.attrs.id,
								t: node.attrs.t,
```

File: src/Socket/messages-recv.ts [hunk 3]
```diff
@@ -99,10 +99,14 @@ export const makeMessagesRecvSocket = (config: SocketConfig) => {
...
 
-		if(tag !== 'message' && attrs.type) {
+		if(!!attrs.type) {
 			stanza.attrs.type = attrs.type
 		}
 
+		if(tag === 'message') {
+			stanza.attrs.from = authState.creds.me!.id
+		}
+
```
Comment (PurpShell) [inline src/Socket/messages-recv.ts:109]:
```suggestion

    if (
      !!attrs.type &&
      (tag !== "message" ||
        getBinaryNodeChild({ tag, attrs, content }, "unavailable"))
    ) {
      stanza.attrs.type = attrs.type;
    }

    if (
      tag === "message" &&
      getBinaryNodeChild({ tag, attrs, content }, "unavailable")
    ) {
      stanza.attrs.from = authState.creds.me!.id;
    }
```

**Review by joaogcamilojr — APPROVED** [review]

**Comment by mfirdausazizi** [comment]
Thanks @PurpShell 

Tested on a freshly registered WhatsApp account and seeing a massive improvement.

Only 1 message failed to be found. Not sure what happened.

`{
    "level": 20,
    "time": "2024-03-13T07:24:43.454Z",
    "pid": 1099062,
    "hostname": "censored",
    "recv": {
        "tag": "notification",
        "attrs": {
            "from": "hidden@s.whatsapp.net",
            "type": "privacy_token",
            "id": "2230837972",
            "t": "1710314682"
        }
    },
    "sent": {
        "id": "2230837972",
        "to": "hidden@s.whatsapp.net",
        "class": "notification",
        "type": "privacy_token"
    },
    "msg": "sent ack"
}
{
    "level": 20,
    "time": "2024-03-13T07:24:43.742Z",
    "pid": 1099062,
    "hostname": "censored",
    "recv": {
        "tag": "notification",
        "attrs": {
            "from": "hidden@s.whatsapp.net",
            "type": "privacy_token",
            "id": "3911202524",
            "t": "1710314683"
        }
    },
    "sent": {
        "id": "3911202524",
        "to": "hidden@s.whatsapp.net",
        "class": "notification",
        "type": "privacy_token"
    },
    "msg": "sent ack"
}
{
    "level": 20,
    "time": "2024-03-13T07:24:43.754Z",
    "pid": 1099062,
    "hostname": "censored",
    "tag": "message",
    "attrs": {
        "from": "hidden@s.whatsapp.net",
        "type": "text",
        "id": "3ABF23C80CCD656C59E1",
        "sender_lid": "102160192766119@lid",
        "notify": "Nurul Syahida Hasbullah",
        "t": "1710314683"
    },
    "content": [
        {
            "tag": "url_text",
            "attrs": {}
        },
        {
            "tag": "url_number",
            "attrs": {}
        },
        {
            "tag": "unavailable",
            "attrs": {}
        }
    ],
    "msg": "missing body; sending ack then ignoring."
}
{
    "level": 20,
    "time": "2024-03-13T07:24:43.754Z",
    "pid": 1099062,
    "hostname": "censored",
    "recv": {
        "tag": "message",
        "attrs": {
            "from": "hidden@s.whatsapp.net",
            "type": "text",
            "id": "3ABF23C80CCD656C59E1",
            "sender_lid": "102160192766119@lid",
            "notify": "Nurul Syahida Hasbullah",
            "t": "1710314683"
        }
    },
    "sent": {
        "id": "3ABF23C80CCD656C59E1",
        "to": "hidden@s.whatsapp.net",
        "class": "message",
        "type": "text",
        "from": "601154452900:1@s.whatsapp.net"
    },
    "msg": "sent ack"
}
{
    "level": 20,
    "time": "2024-03-13T07:24:52.893Z",
    "pid": 1099062,
    "hostname": "censored",
    "recv": {
        "tag": "receipt",
        "attrs": {
            "from": "hidden@s.whatsapp.net",
            "type": "read",
            "id": "3ABF23C80CCD656C59E1",
            "recipient": "hidden@s.whatsapp.net",
            "t": "1710314692"
        }
    },
    "sent": {
        "id": "3ABF23C80CCD656C59E1",
        "to": "hidden@s.whatsapp.net",
        "class": "receipt",
        "recipient": "hidden@s.whatsapp.net",
        "type": "read"
    },
    "msg": "sent ack"
}`

**Comment by PurpShell** [comment]
> Thanks @PurpShell
> 
> Tested on a freshly registered WhatsApp account and seeing a massive improvement.
> 
> Only 1 message failed to be found. Not sure what happened.
> 
> `{ "level": 20, "time": "2024-03-13T07:24:43.454Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "notification", "attrs": { "from": "hidden@s.whatsapp.net", "type": "privacy_token", "id": "2230837972", "t": "1710314682" } }, "sent": { "id": "2230837972", "to": "hidden@s.whatsapp.net", "class": "notification", "type": "privacy_token" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:43.742Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "notification", "attrs": { "from": "hidden@s.whatsapp.net", "type": "privacy_token", "id": "3911202524", "t": "1710314683" } }, "sent": { "id": "3911202524", "to": "hidden@s.whatsapp.net", "class": "notification", "type": "privacy_token" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:43.754Z", "pid": 1099062, "hostname": "censored", "tag": "message", "attrs": { "from": "hidden@s.whatsapp.net", "type": "text", "id": "3ABF23C80CCD656C59E1", "sender_lid": "102160192766119@lid", "notify": "Nurul Syahida Hasbullah", "t": "1710314683" }, "content": [ { "tag": "url_text", "attrs": {} }, { "tag": "url_number", "attrs": {} }, { "tag": "unavailable", "attrs": {} } ], "msg": "missing body; sending ack then ignoring." } { "level": 20, "time": "2024-03-13T07:24:43.754Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "message", "attrs": { "from": "hidden@s.whatsapp.net", "type": "text", "id": "3ABF23C80CCD656C59E1", "sender_lid": "102160192766119@lid", "notify": "Nurul Syahida Hasbullah", "t": "1710314683" } }, "sent": { "id": "3ABF23C80CCD656C59E1", "to": "hidden@s.whatsapp.net", "class": "message", "type": "text", "from": "hidden:1@s.whatsapp.net" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:52.893Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "receipt", "attrs": { "from": "hidden@s.whatsapp.net", "type": "read", "id": "3ABF23C80CCD656C59E1", "recipient": "hidden@s.whatsapp.net", "t": "1710314692" } }, "sent": { "id": "3ABF23C80CCD656C59E1", "to": "hidden@s.whatsapp.net", "class": "receipt", "recipient": "hidden@s.whatsapp.net", "type": "read" }, "msg": "sent ack" }`

Keep in mind there could be some other types of errors occuring other than this one. One I've seen circulate often is the ID one

**Comment by mfirdausazizi** [comment]
> > Thanks @PurpShell
> > Tested on a freshly registered WhatsApp account and seeing a massive improvement.
> > Only 1 message failed to be found. Not sure what happened.
> > `{ "level": 20, "time": "2024-03-13T07:24:43.454Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "notification", "attrs": { "from": "hidden@s.whatsapp.net", "type": "privacy_token", "id": "2230837972", "t": "1710314682" } }, "sent": { "id": "2230837972", "to": "hidden@s.whatsapp.net", "class": "notification", "type": "privacy_token" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:43.742Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "notification", "attrs": { "from": "hidden@s.whatsapp.net", "type": "privacy_token", "id": "3911202524", "t": "1710314683" } }, "sent": { "id": "3911202524", "to": "hidden@s.whatsapp.net", "class": "notification", "type": "privacy_token" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:43.754Z", "pid": 1099062, "hostname": "censored", "tag": "message", "attrs": { "from": "hidden@s.whatsapp.net", "type": "text", "id": "3ABF23C80CCD656C59E1", "sender_lid": "102160192766119@lid", "notify": "Nurul Syahida Hasbullah", "t": "1710314683" }, "content": [ { "tag": "url_text", "attrs": {} }, { "tag": "url_number", "attrs": {} }, { "tag": "unavailable", "attrs": {} } ], "msg": "missing body; sending ack then ignoring." } { "level": 20, "time": "2024-03-13T07:24:43.754Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "message", "attrs": { "from": "hidden@s.whatsapp.net", "type": "text", "id": "3ABF23C80CCD656C59E1", "sender_lid": "102160192766119@lid", "notify": "Nurul Syahida Hasbullah", "t": "1710314683" } }, "sent": { "id": "3ABF23C80CCD656C59E1", "to": "hidden@s.whatsapp.net", "class": "message", "type": "text", "from": "hidden:1@s.whatsapp.net" }, "msg": "sent ack" } { "level": 20, "time": "2024-03-13T07:24:52.893Z", "pid": 1099062, "hostname": "censored", "recv": { "tag": "receipt", "attrs": { "from": "hidden@s.whatsapp.net", "type": "read", "id": "3ABF23C80CCD656C59E1", "recipient": "hidden@s.whatsapp.net", "t": "1710314692" } }, "sent": { "id": "3ABF23C80CCD656C59E1", "to": "hidden@s.whatsapp.net", "class": "receipt", "recipient": "hidden@s.whatsapp.net", "type": "read" }, "msg": "sent ack" }`
> 
> Keep in mind there could be some other types of errors occuring other than this one. One I've seen circulate often is the ID one

I see. Must be a separate error / bug. Thanks for the hard work.

**Review by saifkhan7865 — APPROVED** [review]

**Review by antonnymello — APPROVED** [review]

**Comment by ibinetwork** [comment]
@canove trancou o pull request rsrsrs... libera ai :)

**Review by canove — APPROVED** [review]

**Review by andresayac — APPROVED** [review]

**Review by Fonnte — APPROVED** [review]

**Review by antonnymello — APPROVED** [review]

**Review by vinikjkkj — APPROVED** [review]

**Review by ibinetwork — APPROVED** [review]

**Review by mmhospedagem — APPROVED** [review]

**Pull request merged**